### ISSUE TITLE
Need to define a solution for exporting manifests with more than 65,000 rows

### ISSUE THREAD

**Issue by fedorov** [ImagingDataCommons/IDC-WebApp#394]
65000 limit is too stringent for radiology use cases. A single imaging study can easily have more than 1000 DICOM instances. This was also brought up by Ulli.

Some ideas are below - what is feasible?

* we could reduce the manifest size is to reduce the number of columns
included, and allow user to pick which attributes should be included
in the manifest.

* we also discussed an idea to create a BigQuery table corresponding
to the manifest, and give user access to that table, but this brings
the question about who owns the table, and who would be able to access
that table

* another idea we discussed is giving the user a BigQuery query that
corresponds to the cohort, that they would be able to run against IDC
content and store the result into their tables under their projects.
Given the limitation on the manifest size, it might be that this is
the most expedient solution, and should be prioritized.

I think the most practical one is to give users a BQ query definition, and provide instructions how they can run the query and save results into a table under a project they own.

**Comment by wlongabaugh** [comment]
For option 2, IDC could own the table and only give the user's Google ID read permissions; they can easily copy it to their own project. The table would expire after e.g. 1 week.

**Comment by bcli4d** [comment]
The API (still working on deployment) can return an arbitrarily large manifest, either as a single get, or multiple paged gets. The API calls the web app to initiate the BQ query, but the web app returns only the BQ job reference to the API, which then gets the resulting data from BQ and returns it to the client.
An example shows obtaining just the first 10 rows (page_sze=10) of a manifest of GCS URLs for previously created cohort_id 1443. The manifest has a total of 196844 items.:
bond:temp BillClifford$ curl -X GET "http://localhost:8095/v1/cohorts/1443/manifest?access_method=url&url_access_type=gs&url_region=us&page_size=10" -H "accept: application/json"
```
{
  "code": 200, 
  "cohort": {
    "cohort_id": 1443, 
    "description": "Example description", 
    "filterSet": {
      "filters": {
        "Modality": [
          "CT", 
          "MR"
        ], 
        "collection_id": [
          "TCGA-LUAD", 
          "TCGA-KIRC"
        ], 
        "race": [
          "WHITE"
        ]
      }, 
      "idc_data_version": "1.0"
    }, 
    "name": "mycohort"
  }, 
  "job_reference": {
    "jobId": "e7f46817-c150-42a0-8d56-a06ab9ce763d", 
    "location": "US"
  }, 
  "manifest": {
    "dois": [], 
    "rowsReturned": 10, 
    "totalFound": 196844, 
    "url_access_type": "gs", 
    "url_region": "us", 
    "urls": [
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.100052696407437114116810952334.dcm#1592634365935943", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.100100699868028287673029927535.dcm#1592634365958021", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.100209053339233505710613826900.dcm#1592634365960820", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.100425897261907539607906037105.dcm#1592634366037666", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.102348190012021090074456798132.dcm#1592634366131986", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.102478912827572275729103829122.dcm#1592634366158915", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.102540262711348379845819191909.dcm#1592634366097819", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.103138634118173807147637765421.dcm#1592634366183159", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.103276140140042208458485233305.dcm#1592634366358318", 
      "gs://idc-tcia-tcga-kirc/dicom/1.3.6.1.4.1.14519.5.2.1.1357.4004.191697623679477216024482243437/1.3.6.1.4.1.14519.5.2.1.1357.4004.155696143842342738147540861623/1.3.6.1.4.1.14519.5.2.1.1357.4004.103504183938215203400927901261.dcm#1592634366309145"
    ]
  }, 
  "next_page": "BF63SZJJOUAQAAASA4EAAEEAQCAAKGQEBAFBACRAWCXBK==="
}

```
I currently do not return additional data for each item, such as the associated SOPInstanceUID, it order to make it a little easier to feed the data to gsutil. That could be easily changed. 

The job_reference and next_page values in the returned data can passed in subsequent call to get subsequent pages.

**Comment by fedorov** [comment]
@wlongabaugh sounds good! @s-paquette what would be the complexity of implementing this?

@bcli4d this is great! Would be great if you could add an example notebook exercising this endpoint, once it is ready. Nevertheless, we should address the portal manifest download limitation.

**Comment by s-paquette** [comment]
@wlongabaugh @fedorov This has a caveat; ONLY a google account will work for this. Private IDC accounts won't. So, as long as I can easily identify those individuals--should be doable--this wouldn't be too complex to implement in 2-3 days. We already have BQ export code. This might even be a place to make use of the flattened view, as a query into BQ won't add load to the Solr VM.

**Comment by fedorov** [comment]
@s-paquette as a related issue, I strongly suggest we consider the option to completely remove the non-google login. I think it only adds confusion, and will make maintenance more challenging going forward (handing off cohorts, handling different workflows for google and non-google users). There is very little one can do if they are not willing to create a google account. I think refinement of our login options is a fair game considering we are in the pilot stage.

**Comment by s-paquette** [comment]
@fedorov I'd appreciate an entirely new ticket for this as it changes our security requirements (numerous things are no longer needed if only Google handles our logins).

**Comment by wlongabaugh** [comment]
Here is the Google page describing how to add a user with reader role on a dataset: [https://cloud.google.com/bigquery/docs/dataset-access-controls](https://cloud.google.com/bigquery/docs/dataset-access-controls)

**Comment by s-paquette** [comment]
We can probably just do this instead: https://cloud.google.com/bigquery/docs/table-access-controls

Will need to see if the user gets dataset list access, though.

**Comment by wlongabaugh** [comment]
Smae deal, but for storage: [https://cloud.google.com/storage/docs/access-control/using-iam-permissions#bucket-add](https://cloud.google.com/storage/docs/access-control/using-iam-permissions#bucket-add). Note that this adds users at the bucket level, so we would need to create a bucket for each user manifest download. We could do fine-grained, but I think we can only control this thru gsutil or the XML/JSON API and not using the Python libraries. While we can have Google handle the expiration and deletion of bucket objects, I don't think we can't do that for buckets. But we could have Google expire and delete the manifest after one week, and have a cron job that deletes buckets that are empty and older than a week.

**Comment by wlongabaugh** [comment]
> We can probably just do this instead: https://cloud.google.com/bigquery/docs/table-access-controls
> 
> Will need to see if the user gets dataset list access, though.

True. Though it looks like that feature is not yet in the Python API. Yeah, no good giving them table view access if they can't pin the dataset though.

**Comment by wlongabaugh** [comment]
Also, remember third option is to provide a page of buttons to pull down chunks of the manifest.

**Comment by fedorov** [comment]
I am not sure why a GCS-based solution is relevant here. With the manifest the user needs to have a way to pull the data out into their local resource, not yet another cloud location to deal with. It is also important to be able to preserve in a cost-efficient way the content of a cohort. So I really think we should we focusing on one of the following solutions:
1) create a temporary BQ table corresponding to the manifest
2) give user a query that can be executed against IDC BQ tables to get the subset corresponding to the manifest

**Comment by s-paquette** [comment]
@wlongabaugh I'm not sure what you mean? I see the ability to get and set the IAM policy of a table: 

https://googleapis.github.io/google-api-python-client/docs/dyn/bigquery_v2.tables.html

I thought we were going to expire and delete these, so I'm not sure that being able to pin them would be relevant.

**Comment by fedorov** [comment]
@s-paquette how difficult would it be to provide the query to replicate the cohort (issue #182)? This would not require any additional infrastructure to manage permissions or clean up temporary tables or buckets, and will effectively resolve the issue for the user.

**Comment by s-paquette** [comment]
@fedorov The main issue is that would still require a Google account, and, the query won't be the one which replicates the cohort in terms of counts or the lists in the UI since those are specific to the WebApp's use (and come from Solr anyways). Making a table for the user to access is definitely faster.

**Comment by fedorov** [comment]
@s-paquette won't they need a Google account to access that table anyway?

We could (and probably should) make the view used to populate Solr public anyway. I am not sure I understand what you mean by "the query won't be the one which replicates the cohort in terms of counts or the lists in the UI since those are specific to the WebApp's use".

**Comment by s-paquette** [comment]
@fedorov They will, yes, but if we're giving them query text, we've done almost all of the work to make the table--we might as well make the table. (Put another way, that query isn't current constructed to make the manifest, as the manifest comes off of Solr.)

We will be making that view public, though that view doesn't contain paths for files, as they're much too large. We construct the paths at manifest export time. So the query we give users would be a different query all-together, as it can pull the GCS path from the BQ tables.

**Comment by s-paquette** [comment]
Okay, first pass at BQ user tables is in. This is on dev right now. A couple of notes:

- It turns out giving the user dataViewer on a project they're not in entirely masks the dataset itself. They'll know the name (as it's part of the full canonical table name) but they can't see anything else in the project OR dataset if they're not inheriting those roles otherwise. (Note for you @madelyngreyes you will need to test using a gmail account which is NOT on any of our projects.)
- Users without a GCP project will be dropped into a sandbox project automagically, even without billing set up, if they agree to the main GCP policies dialog box. At least, that's what happened to me.
- For testing purposes, the dev tier's tables are going into the dev tier's project. Practically speaking, this will happen in a special BQ data project. This is just a solution to get this in front of people.

Questions for @fedorov et al:

- BQ table names are vastly more restrictive than file names. Do we want to allow users to supply one, or simply give them one, knowing they have to copy the table immediately anyways?
- Do we want to check for the existence of the same cohort manifest and prevent multiple, chained exports?

**Comment by fedorov** [comment]
> BQ table names are vastly more restrictive than file names. Do we want to allow users to supply one, or simply give them one, knowing they have to copy the table immediately anyways?

I think it is ok to supply one to simplify things.

> Do we want to check for the existence of the same cohort manifest and prevent multiple, chained exports?

Yes, and that's something that I think can be mitigated by adding a link to the already exported table to the cohorts table.

**Comment by s-paquette** [comment]
@fedorov Per the other ticket, if we want to track prior exports (as opposed to just checking for the table)--which is the better, overall approach, because checking the BQ API for extent tables is highly latent--we could do this as a feature enhancement.
